### PR TITLE
Move SDL objects to TLS, allowing multiple windows to co-exist.

### LINKED ISSUE #47: Support for multiple Windows
- Version of embedded-graphics-simulator in use (if applicable): 0.5.0
- Version of embedded-graphics in use (if applicable): latest

## Description of the problem/feature request/other

I have a device with two displays and I wanted to use the simulator as frontend for desktops, and framebuffer for the device itself (removing the `with-sdl` flag when building embedded).

Because of the two displays, I wanted use two `Window` objects, but I of course I got this error:

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: "Cannot initialize `Sdl` more than once at a time."'
```

This doesn't seem to be very hard to fix, at least for my simple use case. There's already a PR (#37). I just wanted to file an issue to track progress of this.

## Test case (if applicable)

```rust
        let window_osd = Window::new(
            "OSD",
            &OutputSettingsBuilder::new()
                .theme(BinaryColorTheme::OledBlue)
                .build(),
        );

        let window_title = Window::new(
            "Title",
            &OutputSettingsBuilder::new()
                .theme(BinaryColorTheme::OledBlue)
                .build(),
        );
```

### PR DESCRIPTION
**Pull request by hansl**
Fixes #47.

Thank you for helping out with embedded-graphics-simulator development! Please:

- [x] Check that you've added passing tests and documentation
- [x] Add a `CHANGELOG.md` entry in the **Unreleased** section under the appropriate heading (**Added**, **Fixed**, etc) if your changes affect the **public API**
- [x] Run `rustfmt` on the project
- [x] Run `just build` (Linux/macOS only) and make sure it passes. If you use Windows, check that CI passes once you've opened the PR.

## PR description

Move SDL objects to TLS, allowing multiple windows to co-exist. This is an updated version of #37.

### CODE REVIEW

**Comment by hansl** [comment]
I have no idea why the `half` dependency would fail for this PR. There might be a moving transitive dependency that update that crate and now it refuses to build in Rust 1.61.

**Comment by hansl** [comment]
@jamwaffles Do you think this is something we could get merged in?

**Comment by pcvonz** [comment]
Would be awesome to get this merged in! For what it's worth, I pulled and re-based this locally and it works great. I have a device which uses both an led matrix and a display and it has been very helpful for testing. @jamwaffles @hansl

File: src/window/sdl_window.rs
Comment (rfuest) [inline src/window/sdl_window.rs]: Isn't this always false? I think one of the variables should be renamed.

**Review by rfuest — CHANGES_REQUESTED** [review]
I can merge the PR, but I'm not sure if I have all necessary permissions to create a new release on crates.io.

**Comment by rfuest** [comment]
Wouldn't polling the events for one window also discard the events for all other windows?

File: src/window/sdl_window.rs
Comment (hansl) [reply src/window/sdl_window.rs]: Oh yeah, you're right. Let me fix.

Comment (hansl) [reply src/window/sdl_window.rs]: Fixed an issue with dropping events if they're for other windows and we don't get to them. Also added an example which can contain an infinite amount of windows.

**Comment by hansl** [comment]
> Wouldn't polling the events for one window also discard the events for all other windows?

I actually saw that too when building the example. In my own code that's not an issue so I missed this. There's now a global event vector for missed events. It's limited and unless someone never polls a window's events it should never be full. I tested it locally at least.

**Review by almindor — APPROVED** [review]
LGTM

File: examples/input-handling-multiple.rs
Comment (rfuest) [inline examples/input-handling-multiple.rs]: I tried to remove the duplicated code to create the window and changed the comments a bit. I would also suggest that we rename this file to `multiple-windows.rs`, because that's the primary functionality it demonstrates. The original name contained `multiple`, but in my opinion it wasn't clear that this refers to windows and not a special type of input handling.
```suggestion
//! # Example: Multiple simulator windows
//!
//! This example demonstrates how multiple simulator windows can be opened at
//! the same time and how each window has it's own event queue. Press `n` to
//! open an additional window. The circle in each window can be moved independently
//! with the arrow keys or by clicking with the mouse.

use embedded_graphics::{
    pixelcolor::Rgb888,
    prelude::*,
    primitives::{Circle, PrimitiveStyle},
};
use embedded_graphics_simulator::{
    sdl2::Keycode, OutputSettings, SimulatorDisplay, SimulatorEvent, Window,
};

const BACKGROUND_COLOR: Rgb888 = Rgb888::BLACK;
const FOREGROUND_COLOR: Rgb888 = Rgb888::RED;
const KEYBOARD_DELTA: i32 = 20;
const CIRCLE_DIAMETER: u32 = 200;

enum LoopResult {
    Continue,
    AddNewWindow,
    Quit,
}

struct ExampleWindow {
    display: SimulatorDisplay<Rgb888>,
    window: Window,
    circle_position: Point,
}

impl ExampleWindow {
    fn new() -> Self {
        Self {
            display: SimulatorDisplay::new(Size::new(800, 480)),
            window: Window::new(
                "Click to move circle (press N for new window)",
                &OutputSettings::default(),
            ),
            circle_position: Point::zero(),
        }
    }

    fn init(&mut self) -> Result<(), core::convert::Infallible> {
        self.circle_position = Point::new(200, 200);

        // Draw the initial circle.
        Circle::with_center(self.circle_position, CIRCLE_DIAMETER)
            .into_styled(PrimitiveStyle::with_fill(FOREGROUND_COLOR))
            .draw(&mut self.display)?;

        Ok(())
    }

    fn handle_events(&mut self) -> Result<LoopResult, core::convert::Infallible> {
        self.window.update(&self.display);

        let events = self.window.events();
        for event in events {
            let new_position = match event {
                SimulatorEvent::Quit => return Ok(LoopResult::Quit),
                SimulatorEvent::KeyDown { keycode, .. } => {
                    let delta = match keycode {
                        Keycode::Left => Point::new(-KEYBOARD_DELTA, 0),
                        Keycode::Right => Point::new(KEYBOARD_DELTA, 0),
                        Keycode::Up => Point::new(0, -KEYBOARD_DELTA),
                        Keycode::Down => Point::new(0, KEYBOARD_DELTA),
                        Keycode::N => return Ok(LoopResult::AddNewWindow),
                        _ => Point::zero(),
                    };

                    Some(self.circle_position + delta)
                }
                SimulatorEvent::MouseButtonUp { point, .. } => Some(point),
                _ => None,
            };

            if let Some(new_position) = new_position {
                // Clear old circle.
                Circle::with_center(self.circle_position, CIRCLE_DIAMETER)
                    .into_styled(PrimitiveStyle::with_fill(BACKGROUND_COLOR))
                    .draw(&mut self.display)?;

                self.circle_position = new_position;

                // Draw circle at new location.
                Circle::with_center(self.circle_position, CIRCLE_DIAMETER)
                    .into_styled(PrimitiveStyle::with_fill(FOREGROUND_COLOR))
                    .draw(&mut self.display)?;
            }
        }
        Ok(LoopResult::Continue)
    }
}

fn main() -> Result<(), core::convert::Infallible> {
    let mut first_window = ExampleWindow::new();
    first_window.init()?;

    let mut windows = vec![first_window];

    'running: loop {
        for window in windows.iter_mut() {
            match window.handle_events()? {
                LoopResult::Continue => {}
                LoopResult::AddNewWindow => {
                    let mut window = ExampleWindow::new();
                    window.init()?;

                    // The `windows` array cannot be modified while it is
                    // mutably borrowed by the `iter_mut` iterator. By breaking
                    // the loop over the windows the lifetime of the iterator is
                    // limited and does no longer interfere with the `push`.
                    windows.push(window);
                    break;
                }
                LoopResult::Quit => {
                    break 'running;
                }
            }
        }
    }

    Ok(())
}
```

**Review by rfuest — CHANGES_REQUESTED** [review]
Adding a example was a good idea.

I've found one bug: It is no longer possible to close the simulator by clicking the close window icon in the titlebar of the window. It still works when one window is displayed, but as soon as at least two windows are open the simulator can only be quit via the `ESC` key.

File: examples/input-handling-multiple.rs
Comment (hansl) [reply examples/input-handling-multiple.rs]: Done, along with a PR to fix the close a window problem.

**Comment by hansl** [comment]
@rfuest PTAL, thanks for the review.

**Comment by rfuest** [comment]
The simulator was always intended to be something to quickly implement a demo and I'm starting to think that adding this feature in this way might no match this original idea.

Here is an experiment I did, which instead let's the user handle the SDL2 initialization and event handling: https://github.com/rfuest/simulator/commit/e2d13142e96f9ec178f8092c0a680ba63297ad1b. In my opinion this keeps the code inside the simulator simpler and is also more flexible for the user, who can handle SDL events any way they like. Please take a look at that code and let me know if this would work for you.

**Comment by hansl** [comment]
@rfuest I don't mind either way. I fixed the problem in my code by basically reimplementing a subset of this library. I just thought more people would like this functionality, as it is fully backward compatible.

**Comment by rfuest** [comment]
> @rfuest I don't mind either way. I fixed the problem in my code by basically reimplementing a subset of this library. I just thought more people would like this functionality, as it is fully backward compatible.

With the addition of the new `SimulatorEvent::WindowEvent` it did technically become a breaking change, although this wouldn't be a problem in most cases, because I don't expect a lot of users will use a `match` expression without a wildcard pattern for the event.

I think that I would prefer my solution, because it also provides advantages for users who don't want to use multiple windows. Not exiting on ESC, which you mentioned in a commend, would be one use case when only one window is used.

Unfortunately I'm not able to open a PR for the code I suggested, because I wouldn't be able to merge it. The permissions are setup in a way where PRs need an approving review from a code owner and I can't review my own code. I'm effectively blocked from writing any code for e-g and the related projects at the moment.

Out of curiosity: Does your application need multiple windows or would it also work by displaying multiple displays in a single window? This had been the way I had intended to implement multi display support in the past. Only having one window that contains multiple displays has the advantage that you can specify the relative positioning in the app, without having to worry about how the window manager places the separate SDL windows. I imagine that creating a simulation for a Stream Deck clone with one display per key would get confusing if you use one window per display. But that's a separate feature and multi windows support is none the less.

**Comment by jamwaffles** [comment]
> The permissions are setup in a way where PRs need an approving review from a code owner and I can't review my own code.

I changed your role from "Member" to "Owner" in the e-g core team group, so you might be able to do more now - if nothing else, to change the setting in repos of your choice to allow merges without reviews. If that doesn't work, I don't know what other levers I have available to pull.

**Comment by rfuest** [comment]
@jamwaffles Thanks for the quick response, but I don't think that did a lot. I can override the branch protection rules for this PR, so maybe I could merge a PR I create in this repo. But I tried to do the same with https://github.com/embedded-graphics/embedded-graphics/pull/738, where I don't have the option to override the branch protection rules. In neither repo I have the permissions to disable the protection rules in the settings.

**Comment by jamwaffles** [comment]
I just made e-g-core-team admin for most repos in the org (was previously "Maintain") which may now allow you to merge e-g/738.

> In neither repo I have the permissions to disable the protection rules in the settings.

The description under "Admin" level implies this is possible, but maybe there's yet another knob I need to twizzle to make it work...

**Comment by rfuest** [comment]
@jamwaffles That has fixed it, thanks!

**Comment by jamwaffles** [comment]
Ah, finally! I've been a terrible steward of this project recently so my apologies for being slow and uncooperative with all this, but it sounds like I don't hold any keys any more which is great :)

**Comment by rfuest** [comment]
> Ah, finally! I've been a terrible steward of this project recently so my apologies for being slow and uncooperative with all this, but it sounds like I don't hold any keys any more which is great :)

Who knows perhaps you will come back in the future to take a more active role in the project again, but until then I'll try to keep the project alive and not mess anything up too much ;).

Would you mind to also make me a co-owner of the GH org? Having at least two owners is recommended by GH: https://docs.github.com/en/organizations/managing-peoples-access-to-your-organization-with-roles/maintaining-ownership-continuity-for-your-organization#about-maintaining-ownership-continuity-for-your-organization

**Comment by jamwaffles** [comment]
> Who knows perhaps you will come back in the future to take a more active role in the project again

I'd still like to get integer-only AA working at some point, so maybe!

> Would you mind to also make me a co-owner of the GH org?

You're an `Owner` now :)

**Comment by rfuest** [comment]
> I'd still like to get integer-only AA working at some point, so maybe!

That would be a nice to have feature. Maybe we'll even have some support for alpha blending by then, which I might work on after the mess that is PR https://github.com/embedded-graphics/embedded-graphics/pull/711 is merged in smaller chunks.

> You're an `Owner` now :)

Thanks again!

**Comment by hansl** [comment]
> Out of curiosity: Does your application need multiple windows or would it also work by displaying multiple displays in a single window?

Multi window is just easier as I have multiple draw buffers (which ends up in different code path on the embedded device). I guess I could technically make it a single display, but for the developer it's easier to see where one display end and the other begins. Just simpler code.

Note that I also duplicated the code because I support sdl3 in my base.

I'll close this PR and if you want to merge this branch into another PR go for it.

**Pull request closed without merging**
